### PR TITLE
HIP-18 Changes for ExpirableTxnRecord

### LINKED ISSUE #1587: Write a custom-fee-aware translator to map from gRPC xfers to full impliedTransfers
**Summary**
Given the custom fee schedules for each token type in a `tokenTransferList`, implement a new `ImpliedTransfers` component that computes the mapping,
```
([TokenType: CustomFees], [TokenTransfers], [ℏTransfers]) -->> ([ℏTransfers], [TokenTransfers]) | errorCode
```

This mapping should be precomputed for each `CryptoTransfer` in `ExpandHandleSpan` using the custom schedules and `GlobalDynamicProperties` available in the working state; if the relevant schedules have not changed upon the `CryptoTransfer` reaching consensus, the output can be used immediately. (And otherwise re-computed.)

### LINKED ISSUE #1590: Update TransactionRecordService to report custom fees
**Summary**
The `ExpirableTxnRecord` resulting from a `CryptoTransfer` that requires custom fee payments will need to report these custom fees.

The `TransactionRecordService` should be able to use the [output of the `ImpliedTransfers` translator](https://github.com/hashgraph/hedera-services/issues/1587) to efficiently set these elements on the `ExpirableTxnRecord` when appropriate.

### LINKED ISSUE #1592: Extend ExpirableTxnRecord with custom fee lists
**Summary**
The `ExpirableTxnRecord` Merkle type must capture the custom fees (ℏ and HTS unit) due for a `CryptoTransfer` involving tokens with custom fee schedules.

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
**Related issue(s)**:
Closes #1590 
Closes #1592 
Closes #1587 

**Summary of the change**:
1. Set Custom fees as `List<CustomFeesBalanceChange>`  in `ExpirableTxnRecord`
2.Modify `ImpliedTransfers` to set CustomFees in two lists. One which is useful for ExpirableTxnRecord `List<CustomFeesBalanceChange> customFeeBalanceChanges` and other with all `CustomFees` for each `EntityId` `List<Pair<EntityId, List<CustomFee>>> customFeesChanges`
3. Add CustomFees to all balance changes in TransferList in `ImpliedTransfersMarshal`
4. Change `BalanceChange` to use `EntityId` instead of `Id` for ease of use
5. Add `CustomFeesBalanceChange.java` SelfSerializable object which is useful to set CustomFees in `ExpirableTxnRecord`
6. Add an interface `CustomFeeSchedules.java` to look up schedules for an Entity
7. Add custom fees to be serialized and deserialized in `ExpirableTxnRecord`
8. Add `FCMCustomFeeSchedules` to look up `List<CustomFee>` in tokens FCMap
9. Recompute the fee schedules from FCMap if any of the fee schedules doesn't match in `wasDerivedFrom`

**External impacts**:
None.

**Applicable documentation**
- [x] Javadoc

### CODE REVIEW

File: hedera-node/src/test/java/com/hedera/services/txns/span/SpanMapManagerTest.java [hunk 1]
```diff
@@ -108,14 +112,14 @@ void recomputesImpliedTransfersIfMetaMatches() {
...
 				.willReturn(someOtherImpliedXfers);
 
 		// when:
 		subject.rationalizeSpan(accessor);
 
 		// then:
-		verify(impliedTransfersMarshal).unmarshalFromGrpc(pretendXferTxn.getCryptoTransfer());
+		verify(impliedTransfersMarshal).unmarshalFromGrpc(pretendXferTxn.getCryptoTransfer(), accessor.getPayer());
 		assertSame(someOtherImpliedXfers, spanMapAccessor.getImpliedTransfers(accessor));
 	}
```
Comment (ljianghedera) [inline hedera-node/src/test/java/com/hedera/services/txns/span/SpanMapManagerTest.java:124]: Seems like we need one scenario here: dynamicProperties stays the same, but a tokenUpdate changes its custom fee schedule, which shall trigger a re-computation of the ImpliedTransfers.

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java [hunk 2]
```diff
@@ -66,17 +74,117 @@ public ImpliedTransfers unmarshalFromGrpc(CryptoTransferTransactionBody op) {
...
+				}
+			}
+
+			List<CustomFee> customFeesOfToken = customFeeSchedules.lookupScheduleFor(scopingToken);
+			entityCustomFees.add(new Pair<>(scopingToken, customFeesOfToken));
+			List<BalanceChange> customFeeChanges = computeBalanceChangeForCustomFee(scopingToken, payerId, amount,
+					customFeesOfToken);
+			changes.addAll(customFeeChanges);
+			customFeeBalanceChangesForRecord.addAll(getListOfBalanceChangesForCustomFees(customFeeChanges));
+		}
```
Comment (ljianghedera) [inline hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java:100]: One question regarding the custom fee change calculation. Here is the scenario: A transfer will transfer token A and B to other accounts. A has its custom fee schedules which require to transfer token C to some account. C again has its custom fee schedule.

The question is, shall we calculate the balance changes recursively?

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java [hunk 3]
```diff
@@ -66,17 +74,117 @@ public ImpliedTransfers unmarshalFromGrpc(CryptoTransferTransactionBody op) {
...
 		}
+		var payerId = EntityId.fromGrpcAccountId(payer);
 		for (var scopedTransfers : op.getTokenTransfersList()) {
 			final var grpcTokenId = scopedTransfers.getToken();
-			final var scopingToken = Id.fromGrpcToken(grpcTokenId);
+			final var scopingToken = EntityId.fromGrpcTokenId(grpcTokenId);
+			var amount = 0L;
 			for (var aa : scopedTransfers.getTransfersList()) {
 				changes.add(tokenAdjust(scopingToken, grpcTokenId, aa));
+				if (aa.getAmount() > 0) {
```
Comment (ljianghedera) [inline hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java:89]: Should we also consider the amount transfer in to the payer account? For a hypothetical scenario, the payer initially has 100 A token. In a transfer transaction that will transfer 200 A token from account X to payer, and transfer 250 A token to account Y. From the payer's view point, the transaction should go through. But the current logic may report insufficient balance error.

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java (hunk 2 above)
Comment (qnswirlds) [reply hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java:100]: This is a good question @ljianghedera. I think we should do only simple transfers and charges. A mutually recursive transfer (transfer of token A requires custom fee of B and transfer of token B also requires custom fee of A), if not being handled as mentioned, will complicate a lot in the fee calculations and also in our understandings/interpretations.

Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java:100]: Yeah, it is. So I think we need to stick to current implementation, and we may need to document this to address potential confusions and questions.

**Review by ljianghedera — DISMISSED** [review]
Overall LGTM except the minor comments.

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfers.java [hunk 4]
```diff
@@ -41,38 +45,54 @@
 public class ImpliedTransfers {
 	private final ImpliedTransfersMeta meta;
 	private final List<BalanceChange> changes;
+	private final List<Pair<EntityId, List<CustomFee>>> entityCustomFees;
+	private final List<CustomFeesBalanceChange> customFeesBalanceChanges;
 
-	private ImpliedTransfers(ImpliedTransfersMeta meta, List<BalanceChange> changes) {
+	private ImpliedTransfers(ImpliedTransfersMeta meta, List<BalanceChange> changes,
+			List<Pair<EntityId, List<CustomFee>>> entityCustomFees,
+			List<CustomFeesBalanceChange> customFeeBalanceChanges) {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfers.java:53]: Related to my comment on the protobuf [here](https://github.com/hashgraph/hedera-protobufs/pull/42/files#r657356764), imo we should rename `CustomFeesBalanceChange` to `AssessedCustomFee`.

So `ImpliedTransfers` would have a `List<AssessedCustomFee> assessedCustomFees` field that gives all the custom fees that must be paid for the `CryptoTransfer` to succeed.

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfers.java [hunk 5]
```diff
@@ -41,38 +45,54 @@
 public class ImpliedTransfers {
 	private final ImpliedTransfersMeta meta;
 	private final List<BalanceChange> changes;
+	private final List<Pair<EntityId, List<CustomFee>>> entityCustomFees;
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfers.java:48]: Something like `involvedTokenFeeSchedules` might be more readable here.

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java:96]: Great point! This is addressed by consolidating all `(account, token/ℏ)` balance changes via the `existingBalanceChanges` map.

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java [hunk 6]
```diff
@@ -66,17 +76,156 @@ public ImpliedTransfers unmarshalFromGrpc(CryptoTransferTransactionBody op) {
...
+	 */
+	private List<BalanceChange> computeBalanceChangeForCustomFee(EntityId scopingToken,
+			EntityId payerId,
+			long totalAmount, List<CustomFee> customFeesOfToken,
+			Map<Pair<EntityId, EntityId>, BalanceChange> existingBalanceChanges,
+			List<CustomFeesBalanceChange> customFeeBalanceChangesForRecord) {
+		List<BalanceChange> customFeeChanges = new ArrayList<>();
+		for (CustomFee fees : customFeesOfToken) {
+			if (fees.getFeeType() == CustomFee.FeeType.FIXED_FEE) {
+				addFixedFeeBalanceChanges(fees, payerId, customFeeChanges, existingBalanceChanges,
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java:122]: Nice. 😁

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleToken.java [hunk 7]
```diff
@@ -490,8 +490,11 @@ public void setFeeScheduleFrom(CustomFeesOuterClass.CustomFees grpcFeeSchedule)
 		return	grpcFeeSchedule.getCustomFeesList().stream().map(CustomFee::fromGrpc).collect(toList());
 	}
 
-	public List<CustomFee> customFeeSchedule() {
+	public List<CustomFee> getFeeSchedule() {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleToken.java:493]: Looks like we have duplicate getters here.

File: hedera-node/src/main/java/com/hedera/services/state/submerkle/CustomFeesBalanceChange.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import proto.CustomFeesOuterClass;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Process self serializable object that encapsulates a balance change, either ℏ or token unit.
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/state/submerkle/CustomFeesBalanceChange.java:37]: Along with the rename to `AssessedCustomFee`, would emphasize that this type represents an **assessed** custom fee, which may or may not result in a balance change, depending on if the payer can afford it.

**Review by tinker-michaelj — COMMENTED** [review]
Awesome work! 👏 

Aside from the rename `CustomFeeBalanceChange -->> AssessedCustomFee`, which does seem to clarify semantics quite a bit, this all looks good to me!!

File: hedera-node/src/test/java/com/hedera/services/txns/span/SpanMapManagerTest.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/test/java/com/hedera/services/txns/span/SpanMapManagerTest.java:144]: Added. Thanks @ljianghedera

File: hedera-node/src/main/java/com/hedera/services/state/submerkle/CustomFeesBalanceChange.java (hunk 8 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/state/submerkle/CustomFeesBalanceChange.java]: Changed javadoc. Thanks @tinker-michaelj

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleToken.java (hunk 7 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleToken.java]: Oh yeah .. I guess TokenCreate PR has added it. Removed duplicated getter.

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfers.java (hunk 5 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfers.java]: Fixed !

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfers.java (hunk 4 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfers.java]: Renamed to `AssessedCustomFee` . Thanks @tinker-michaelj

**Review by ljianghedera — APPROVED** [review]
Looks great!

**Review by tinker-michaelj — APPROVED** [review]
LGTM!!

**Review by qnswirlds — APPROVED** [review]
Some new lines are reported not covered by tests. But I think we can improve that later. Overall LGTM.

**Pull request merged**